### DISCUSSION TITLE
Skia Backend

### DISCUSSION THREAD

**Discussion by KnIfER** [alandefreitas/matplotplusplus#190] [Ideas]
Skia is more flexibel. Skia itself is backed by cpu, opengl, dx ....
The code for skia backend is shorter and more readable.

skia backend / draw_path
```
void draw_path(const std::vector<double> &x,
	const std::vector<double> &y,
	const std::array<float, 4> &color) override {

	SkPaint* stroke = new SkPaint();
	stroke->setColor(SkColorSetARGB((1.f - color[0])*255, (color[1])*255, (color[2])*255, (color[3])*255));
	stroke->setAntiAlias(true);
	stroke->setStroke(true);
	stroke->setStrokeWidth(2.0f);

	SkPathBuilder* path_builder = new SkPathBuilder();
	int length = std::min(x.size(), y.size()); // , color.size()
	for (size_t i = 0; i < length; i++)
	{
		if (i==0)
		{
			path_builder->moveTo (x[i], y[i]);
		}
		else
		{
			path_builder->lineTo (x[i], y[i]);
		}
	}
	SkPath path = path_builder->detach();
	_canvas->drawPath(path, *stroke);

	delete path_builder;
	delete stroke;
}
```
![image](https://user-images.githubusercontent.com/11593903/132298767-1736bdcc-fe4b-49f7-8738-f5664ee33a68.png)

**Comment by alandefreitas** [comment]
Hi @KnIfER !

Skia is indeed a great option. I remember considering Skia and Agg in the past, but I think I discarded these ideas because they were 2D only and Agg is not being maintained. No one has ever brought it up since then. I imagine it's even more readable if we declare `SkPaint` and `SkPathBuilder` with scoped unique pointers.

In retrospect, this might have been a bad decision, since the 3D plots we need are quite simple and we can calculate the projection matrices trivially with GLM. In that case, the backends would only need to implement the pure virtual functions for 2D drawing in the interface. The 3D drawing would be delegated to GLM unless they are overridden. 

That would make new backends much easier to implement. For instance, WxWidgets, Qt, ImGUI have some drawing functions that look like Skia's. I don't think there's any problem allowing a Skia backend (BSD-3) with Matplot++ (MIT) either unless we are modifying Skia, but I'd have to research a little more to be sure.

I imagine you have some experience with Skia, right? It should be trivial to replicate the experimental OpenGL 2D line plot with Skia as you did. That would be a great starting point, as all that's missing in that example is text rendering, which should be easy with Skia. If you are interested in getting this started, I'm in.

**Comment by KnIfER** [comment]
The test code for the line plot is here : https://github.com/KnIfER/Textrument-Plugins/blob/main/DirectUILib/SkiaPlotLib/SKiaPlotTest/TestSkiaBackend.cpp

You can mix opengl 3d with skia 2d graphics:
https://github.com/KnIfER/Textrument-Plugins/blob/main/DirectUILib/SkiaTests/GLSkiaHelloWorld.cpp
![image (1)](https://user-images.githubusercontent.com/11593903/132935350-d9774e55-3e3e-4cf4-b9d3-a67a468878dd.png)

**Reply by alandefreitas** [reply]
That looks really good. If anyone is interested in leading that, I'm all in.
